src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,399 @@
+you can see what nullability the tool infers.
+For example, in the preceding screenshot,
+the tool infers that the `bar2()` function returns a non-nullable type (`int`).
+
+{{ site.alert.tip }}
+  To see the reason for each change (or non-change),
+  click the type.
+{{ site.alert.end }}
+
+When analysis infers the wrong nullability,
+you can override its choice by inserting temporary `/*?*/` and `/*!*/` markers.
+
+* In the tool UI, you can insert these markers using the
+  **Add `/*?*/` hint** and **Add `/*!*/` hint** buttons.
+* Alternatively, switch back to your editor and change the code yourself,
+  even while the tool is still running.
+  When you've finished editing your code,
+  click **Rerun from sources** to pick up your changes.
+
+{{ site.alert.important }}
+  Don't use new null safety features yet.
+  Your code is still null-unsafe for now.
+{{ site.alert.end }}
+
+When you like all of the changes
+that the migration tool proposes &mdash; all of the
+`?`, `required`, `!`, and other changes &mdash; click
+**Apply migration** to tell the tool to save the changes.
+Congratulations, your package is migrated!
+
+
+### Migrating by hand
+
+Manual migration can be the right choice if you
+want to migrate your package incrementally.
+Because Dart supports [mixed-version programs][],
+you can migrate one library (generally one Dart file) at a time,
+while still being able to run your app and its tests.
+
+[mixed-version programs]: /go/unsound-null-safety
+
+We recommend that you **first migrate leaf libraries** —
+libraries that don't import other files from the package.
+Then migrate libraries that directly depend on the leaf libraries.
+End by migrating the libraries that have the most
+intra-package dependencies.
+
+For example, say you have a `lib/src/util.dart` file
+that imports other (null-safe) packages and core libraries,
+but that doesn't have any `import '<local_path>'` directives.
+Consider migrating `util.dart` first,
+and then migrating simple files that depend only on `util.dart`.
+If any libraries have cyclic imports
+(for example, A imports B which imports C, and C imports A),
+consider migrating those libraries together.
+
+To migrate a package by hand, follow these steps:
+
+1. Modify the SDK constraints in the package's `pubspec.yaml` file,
+   setting the language version to 2.12 beta or later:
+   ```yaml
+environment:
+  sdk: '>=2.12.0-18.0.beta <3.0.0' [PENDING: check version; would 2.12.0-0 work?]
+```
+
+2. Update dependencies: <br>
+   ```terminal
+$ dart pub get
+```
+
+   Running `dart pub get` with a lower constraint of 2.12.0
+   sets the default language version of
+   every library in your package to 2.12,
+   opting them all in to null safety.
+
+
+3. Open the package in your IDE. <br>
+   You're likely to see a lot of analysis errors.
+   That's OK.
+
+4. _Optional:_ Add a [language version comment][] to the top of
+   any Dart files that you don't want to consider during your current migration:
```

Comment:
I wonder if this should be omitted here, and then all mentioning of unsound null safety & mixed version should be elsewhere?

Reply:
As long as it IS elsewhere. This is going to mess up a bunch of folks.

Reply:
The top of this section motivates the whole migrating by hand thing using incremental migration, and it mentions mixed-version programs. So... I'm not sure that deleting this step would be helpful/enough.

Reply:
What I meant was to potentially move the whole "Migrating by hand" section to a new "Advanced manual null safety" migration topic, which could also hold https://kw-www-dartlang-1.firebaseapp.com/guides/language/evolution#per-library-language-version-selection. It would simplify the flow here, which is already complicated.

Reply:
As discussed, we changed out mind and would like to keep "hand migration" here (but can still move mixed-version elsewhere). Resolving.